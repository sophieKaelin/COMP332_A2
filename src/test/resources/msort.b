//
// This file is part of COMP332 Assignment 2/3 2019.
//
// weBCPL, a retro BCPL to WebAssembly compiler.
//
// © 2019, Dominic Verity and Anthony Sloane, Macquarie University.
//         All rights reserved.
// © 2005, Martin Richards
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// Sample sorting code from Martin Richards BCPL distribution.
//

MANIFEST $( upb=60; big=1000  $)

LET start() BE
$(
    LET p, q = 1, 0
    LET v = VEC upb
    AND w = VEC upb
    FOR i = 1 TO upb-1 DO v!i := 'A' + randno(26) - 1
    v!0, v!upb := big, big
    $(
        q := 0
        FOR i = 1 TO p-1 DO wrch(v!i)
        wrch('|')
        FOR i = p TO upb-1 DO $( q:=q+1; w!q := i; wrch(v!i) $)
        newline()
        q := q+1
        UNTIL q<=2 DO
        $(
            LET r = 0
            w!q := upb
            FOR i = 1 TO q-1 DO
                IF v!(w!(i-1))>v!(w!i)<=v!(w!(i+1)) DO $( r:=r+1; w!r:=w!i $)
            q := r+1
            w!q := upb
            FOR i = 1 TO p-1 DO wrch(v!i)
            wrch('|')
            r := 1
            FOR i = p TO upb-1 TEST w!r=i
                               THEN $( wrch(v!i); r := r+1 $)
                               ELSE wrch(' ')
            newline()
        $)
        q := v!p
        v!p := v!(w!1)
        v!(w!1) := q
        p := p+1
    $) REPEATUNTIL p>=upb
$)