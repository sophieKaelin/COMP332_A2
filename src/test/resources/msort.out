Program(
    Vector(
        ManifestDecl(
            Vector(
                ManifestEntry(IdnDef("upb"), Some(IntExp(60))),
                ManifestEntry(IdnDef("big"), Some(IntExp(1000))))),
        LetDecl(
            Vector(
                LetProcClause(
                    IdnDef("start"),
                    Vector(),
                    Block(
                        Vector(
                            LetDecl(
                                Vector(
                                    LetVarClause(
                                        Vector(IdnDef("p"), IdnDef("q")),
                                        Vector(IntExp(1), IntExp(0))))),
                            LetDecl(
                                Vector(
                                    LetVecClause(IdnDef("v"), IdnExp(IdnUse("upb"))),
                                    LetVecClause(IdnDef("w"), IdnExp(IdnUse("upb")))))),
                        Vector(
                            ForStmt(
                                IdnDef("i"),
                                IntExp(1),
                                MinusExp(IdnExp(IdnUse("upb")), IntExp(1)),
                                None,
                                AssignStmt(
                                    Vector(BinaryPlingExp(IdnExp(IdnUse("v")), IdnExp(IdnUse("i")))),
                                    Vector(
                                        MinusExp(
                                            PlusExp(
                                                ChrExp(Vector(65)),
                                                CallExp(IdnExp(IdnUse("randno")), Vector(IntExp(26)))),
                                            IntExp(1))))),
                            AssignStmt(
                                Vector(
                                    BinaryPlingExp(IdnExp(IdnUse("v")), IntExp(0)),
                                    BinaryPlingExp(IdnExp(IdnUse("v")), IdnExp(IdnUse("upb")))),
                                Vector(IdnExp(IdnUse("big")), IdnExp(IdnUse("big")))),
                            RepeatUntilStmt(
                                Block(
                                    Vector(),
                                    Vector(
                                        AssignStmt(Vector(IdnExp(IdnUse("q"))), Vector(IntExp(0))),
                                        ForStmt(
                                            IdnDef("i"),
                                            IntExp(1),
                                            MinusExp(IdnExp(IdnUse("p")), IntExp(1)),
                                            None,
                                            CallStmt(
                                                CallExp(
                                                    IdnExp(IdnUse("wrch")),
                                                    Vector(
                                                        BinaryPlingExp(
                                                            IdnExp(IdnUse("v")),
                                                            IdnExp(IdnUse("i"))))))),
                                        CallStmt(
                                            CallExp(
                                                IdnExp(IdnUse("wrch")),
                                                Vector(ChrExp(Vector(124))))),
                                        ForStmt(
                                            IdnDef("i"),
                                            IdnExp(IdnUse("p")),
                                            MinusExp(IdnExp(IdnUse("upb")), IntExp(1)),
                                            None,
                                            Block(
                                                Vector(),
                                                Vector(
                                                    AssignStmt(
                                                        Vector(IdnExp(IdnUse("q"))),
                                                        Vector(
                                                            PlusExp(IdnExp(IdnUse("q")), IntExp(1)))),
                                                    AssignStmt(
                                                        Vector(
                                                            BinaryPlingExp(
                                                                IdnExp(IdnUse("w")),
                                                                IdnExp(IdnUse("q")))),
                                                        Vector(IdnExp(IdnUse("i")))),
                                                    CallStmt(
                                                        CallExp(
                                                            IdnExp(IdnUse("wrch")),
                                                            Vector(
                                                                BinaryPlingExp(
                                                                    IdnExp(IdnUse("v")),
                                                                    IdnExp(IdnUse("i"))))))))),
                                        CallStmt(CallExp(IdnExp(IdnUse("newline")), Vector())),
                                        AssignStmt(
                                            Vector(IdnExp(IdnUse("q"))),
                                            Vector(PlusExp(IdnExp(IdnUse("q")), IntExp(1)))),
                                        UntilDoStmt(
                                            LessOrEqualExp(IdnExp(IdnUse("q")), IntExp(2)),
                                            Block(
                                                Vector(
                                                    LetDecl(
                                                        Vector(
                                                            LetVarClause(
                                                                Vector(IdnDef("r")),
                                                                Vector(IntExp(0)))))),
                                                Vector(
                                                    AssignStmt(
                                                        Vector(
                                                            BinaryPlingExp(
                                                                IdnExp(IdnUse("w")),
                                                                IdnExp(IdnUse("q")))),
                                                        Vector(IdnExp(IdnUse("upb")))),
                                                    ForStmt(
                                                        IdnDef("i"),
                                                        IntExp(1),
                                                        MinusExp(IdnExp(IdnUse("q")), IntExp(1)),
                                                        None,
                                                        IfDoStmt(
                                                            AndExp(
                                                                GreaterExp(
                                                                    BinaryPlingExp(
                                                                        IdnExp(IdnUse("v")),
                                                                        BinaryPlingExp(
                                                                            IdnExp(IdnUse("w")),
                                                                            MinusExp(
                                                                                IdnExp(IdnUse("i")),
                                                                                IntExp(1)))),
                                                                    BinaryPlingExp(
                                                                        IdnExp(IdnUse("v")),
                                                                        BinaryPlingExp(
                                                                            IdnExp(IdnUse("w")),
                                                                            IdnExp(IdnUse("i"))))),
                                                                LessOrEqualExp(
                                                                    BinaryPlingExp(
                                                                        IdnExp(IdnUse("v")),
                                                                        BinaryPlingExp(
                                                                            IdnExp(IdnUse("w")),
                                                                            IdnExp(IdnUse("i")))),
                                                                    BinaryPlingExp(
                                                                        IdnExp(IdnUse("v")),
                                                                        BinaryPlingExp(
                                                                            IdnExp(IdnUse("w")),
                                                                            PlusExp(
                                                                                IdnExp(IdnUse("i")),
                                                                                IntExp(1)))))),
                                                            Block(
                                                                Vector(),
                                                                Vector(
                                                                    AssignStmt(
                                                                        Vector(IdnExp(IdnUse("r"))),
                                                                        Vector(
                                                                            PlusExp(
                                                                                IdnExp(IdnUse("r")),
                                                                                IntExp(1)))),
                                                                    AssignStmt(
                                                                        Vector(
                                                                            BinaryPlingExp(
                                                                                IdnExp(IdnUse("w")),
                                                                                IdnExp(IdnUse("r")))),
                                                                        Vector(
                                                                            BinaryPlingExp(
                                                                                IdnExp(IdnUse("w")),
                                                                                IdnExp(IdnUse("i"))))))))),
                                                    AssignStmt(
                                                        Vector(IdnExp(IdnUse("q"))),
                                                        Vector(
                                                            PlusExp(IdnExp(IdnUse("r")), IntExp(1)))),
                                                    AssignStmt(
                                                        Vector(
                                                            BinaryPlingExp(
                                                                IdnExp(IdnUse("w")),
                                                                IdnExp(IdnUse("q")))),
                                                        Vector(IdnExp(IdnUse("upb")))),
                                                    ForStmt(
                                                        IdnDef("i"),
                                                        IntExp(1),
                                                        MinusExp(IdnExp(IdnUse("p")), IntExp(1)),
                                                        None,
                                                        CallStmt(
                                                            CallExp(
                                                                IdnExp(IdnUse("wrch")),
                                                                Vector(
                                                                    BinaryPlingExp(
                                                                        IdnExp(IdnUse("v")),
                                                                        IdnExp(IdnUse("i"))))))),
                                                    CallStmt(
                                                        CallExp(
                                                            IdnExp(IdnUse("wrch")),
                                                            Vector(ChrExp(Vector(124))))),
                                                    AssignStmt(
                                                        Vector(IdnExp(IdnUse("r"))),
                                                        Vector(IntExp(1))),
                                                    ForStmt(
                                                        IdnDef("i"),
                                                        IdnExp(IdnUse("p")),
                                                        MinusExp(IdnExp(IdnUse("upb")), IntExp(1)),
                                                        None,
                                                        TestThenElseStmt(
                                                            EqualExp(
                                                                BinaryPlingExp(
                                                                    IdnExp(IdnUse("w")),
                                                                    IdnExp(IdnUse("r"))),
                                                                IdnExp(IdnUse("i"))),
                                                            Block(
                                                                Vector(),
                                                                Vector(
                                                                    CallStmt(
                                                                        CallExp(
                                                                            IdnExp(IdnUse("wrch")),
                                                                            Vector(
                                                                                BinaryPlingExp(
                                                                                    IdnExp(
                                                                                        IdnUse("v")),
                                                                                    IdnExp(
                                                                                        IdnUse("i")))))),
                                                                    AssignStmt(
                                                                        Vector(IdnExp(IdnUse("r"))),
                                                                        Vector(
                                                                            PlusExp(
                                                                                IdnExp(IdnUse("r")),
                                                                                IntExp(1)))))),
                                                            CallStmt(
                                                                CallExp(
                                                                    IdnExp(IdnUse("wrch")),
                                                                    Vector(ChrExp(Vector(32))))))),
                                                    CallStmt(
                                                        CallExp(IdnExp(IdnUse("newline")), Vector()))))),
                                        AssignStmt(
                                            Vector(IdnExp(IdnUse("q"))),
                                            Vector(
                                                BinaryPlingExp(
                                                    IdnExp(IdnUse("v")),
                                                    IdnExp(IdnUse("p"))))),
                                        AssignStmt(
                                            Vector(
                                                BinaryPlingExp(
                                                    IdnExp(IdnUse("v")),
                                                    IdnExp(IdnUse("p")))),
                                            Vector(
                                                BinaryPlingExp(
                                                    IdnExp(IdnUse("v")),
                                                    BinaryPlingExp(IdnExp(IdnUse("w")), IntExp(1))))),
                                        AssignStmt(
                                            Vector(
                                                BinaryPlingExp(
                                                    IdnExp(IdnUse("v")),
                                                    BinaryPlingExp(IdnExp(IdnUse("w")), IntExp(1)))),
                                            Vector(IdnExp(IdnUse("q")))),
                                        AssignStmt(
                                            Vector(IdnExp(IdnUse("p"))),
                                            Vector(PlusExp(IdnExp(IdnUse("p")), IntExp(1)))))),
                                GreaterOrEqualExp(IdnExp(IdnUse("p")), IdnExp(IdnUse("upb")))))))))))
